chase/frontend/components/dashboard/timer.tsx

```diff
@@ -0,0 +1,120 @@
+import React, { useEffect, useRef } from "react";
+import WidgetTemplate from "./widgettemplate";
+import { Toast } from "primereact/toast";
+
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+import { faGavel } from "@fortawesome/free-solid-svg-icons";
+import { faComments } from "@fortawesome/free-solid-svg-icons";
+import { faCoffee } from "@fortawesome/free-solid-svg-icons";
+import { faCirclePause } from "@fortawesome/free-solid-svg-icons";
+
+export default function TimerWidget({
+  headline,
+  until,
+  category,
+}: {
+  headline: string;
+  until: Date | null;
+  category: "formal" | "informal" | "pause" | "suspension";
+}) {
+  const toast = useRef(null);
+
+  const showTimerToast = () => {
+    toast.current.show({
+      severity: "info",
+      summary: "Zeit abgelaufen",
+      detail: "Rückkehr zur formellen Sitzung",
+      sticky: true,
+    });
+  };
+
+  const timeStamp = until?.toLocaleTimeString("de-DE", {
+    hour: "2-digit",
+    minute: "2-digit",
+  });
+
+  const styles = () => {
+    switch (category) {
+      case "formal":
+        return "";
+      case "informal":
+        return "bg-red-500 text-white";
+      case "pause":
+        return "bg-secondary text-white";
+      case "suspension":
+        return "bg-gray-700 text-white";
+    }
+  };
+
+  return (
+    <>
+      <WidgetTemplate cardTitle="" styles={styles()}>
+        <Toast ref={toast} position="bottom-center" />
```

Comment:
This is also something we could put in the layout of the whole app and the reuse across all components

Reply:
Okay! But how exactly do I do that? How do I reference it in the widget? I think I have no full understanding of useRef yet... Can you please suggest an example to see how its done?

Reply:
Sorry, do you mean the Widget Template or the Toast?

In case of the Toast: How exactly do I do that? How do I reference it in the widget? I think I have no full understanding of useRef yet... Can you please suggest an example to see how its done?

Reply:
Unfortunately I'm not an expert myself but the general idea is to have a toast component in the main layout of the app. In our case this would be `chase\frontend\app\layout.tsx` and then reference that throughout the whole app.

Passing the reference through the application would be a task for our yet to establish state management solution. Most likely an app wide react context.

Reply:
@TheXires Do you know about a workflow for this?

Reply:
You can create a context and wrap it around the base component of the entire page. So that it is then available everywhere and can be called via a function. 

Here is an example of how the context might look in this case.

``` jsx
import React, { createContext, useRef } from 'react';
import { Toast, ToastMessage } from 'primereact/toast';

export const ToastContext = createContext({} as ToastContextType);

export const ToastProvider = ({ children }: { children: React.ReactNode }) => {
  const toast = useRef<Toast>(null);

  const showTimerToast = (message: ToastMessage | ToastMessage[]) => {
    toast.current?.show(message);
  };

  return (
    <ToastContext.Provider value={{ showTimerToast }}>
      <Toast ref={toast} position="bottom-center" />
      {children}
    </ToastContext.Provider>
  );
};

export interface ToastContextType {
  showTimerToast: (message: ToastMessage | ToastMessage[]) => void;
}

```

`message` can be customized here of course, then only the types must be adapted accordingly.

Reply:
@TheXires Thanks! I just implemented it. It works – but can you please quickly glance over it, so that I know if I implemented it correctly? :D

Reply:
@TadeSF Yes, that looks pretty good so far. I would only have two small comments.

This code is a context that surrounds the application and not a component that is used in the code. Therefore I would suggest to create a new folder contexts on the same level as components and app.

I would also collect the types all in one place. This way it's easier to find out if a type already exists or needs to be created than if you have to search the entire code base for it.

And is there a reason why you created an own type ToastMessage? If not I would just go with the default on, because then there is no chance of confusion. At the calling point you can then fix Typescript errors by specifying the type explicitly.

``` jsx
const showTimerToast = () => {
  const message: ToastMessage = {
    summary: 'Summary',
    detail: 'Details',
    severity: 'info',
    sticky: true,
  };
  showToast(message);
};
```